### PR TITLE
feat: profile pages

### LINKED ISSUE #23: Add a profile page on desktop
As a user, I want to have a dedicated profile site on desktop, so that I can easily access/edit my provided informations and so that other users can easily identify me.

For Mock-Up scroll down

### Dependencies

- #19 

### Acceptance Criteria

- [ ] a site solely dedicated to my profile (path should be something like "[url.de]/user/me" or maybe something like HPI-Moodle "https://moodle.hpi.de/user/profile.php?id=1216")
- [ ] the style matches the theme of HPI/ Mock-Ups
- [ ] my full name is visible
- [ ] I can edit my name
- [ ] my HPI-Email is visible

Desktop Mock-Up:
profile page desktop

### LINKED ISSUE #26: Add description to profile pages
As a user, I want to be able, to add my on description to my profile and see the description of others (if provided), so that information which does not have an own category/fiels (e.g. name, phone number) can be easily communicated.

### Dependencies

- #23 

### Definition of Done

- [ ] I can add my own description  to the profile page
- [ ] the description is then displayed on my profile page

### LINKED ISSUE #27: Add ability to see the profile pages of others
As a user, I want to be able, to see the profile pages of others, so that I can access relevant informations about users I want to interact with.

### Dependencies

- #23 

### Definition of Done

- [ ] I can see the profile pages of others (path should be something like [url.de]/user/[identifier for user], maybe something like HPI-Moodle "https://moodle.hpi.de/user/profile.php?id=1216")
- [ ] I cannot edit other users information

### LINKED ISSUE #42: Add a profile page on mobile
As a user, I want to have a dedicated profile site on mobile, so that I can easily access/edit my provided informations and so that other users can easily identify me.

For Mock-Up scroll down

### Dependencies

- #19 

### Definition of Done

- [ ] a site solely dedicated to my profile (path should be something like "[url.de]/user/me" or maybe something like HPI-Moodle "https://moodle.hpi.de/user/profile.php?id=1216")
- [ ] the style matches the theme of HPI/ Mock-Ups
- [ ] my full name is visible
- [ ] ~I can edit my name~
- [ ] my HPI-Email is visible

Mobile Mock-Up:

Image

### PR DESCRIPTION
**Pull request by Skn0tt**
Closes #42
Closes #23
Closes #26
Closes #27

### CODE REVIEW

File: spec/features/users/show_spec.rb
Comment (Skn0tt) [inline spec/features/users/show_spec.rb:31]: I think we should assert that it's showing the second user's username here.

File: spec/requests/profiles_spec.rb
Comment (Skn0tt) [inline spec/requests/profiles_spec.rb:5]: @Shieka @phkeese this prints a warning: https://github.com/hpi-swt2/bookkeeper-portal-red/actions/runs/3480448631/jobs/5820258887#step:5:35

I'll go ahead and delete the file, since we already tests for this in spec/features/users/show_spec.rb.

File: app/controllers/profiles_controller.rb
Comment (luap42) [inline app/controllers/profiles_controller.rb:7]: Redundant: `current_user` is equal to `User.find(current_user.id)`.

File: app/controllers/profiles_controller.rb
Comment (luap42) [inline app/controllers/profiles_controller.rb:8]: I'd prefer not to have two different but exactly equal pages for one content (`/profiles/me` and `/profiles/<my-user-id>`), but instead suggest to use redirects to lead the client.

Example:

```
def show_me
    redirect_to profile_path(current_user)
end
```

File: app/views/devise/registrations/edit.html.erb
Comment (luap42) [inline app/views/devise/registrations/edit.html.erb:34]: Suggestion: when this page is currently updated anyway, it may be a good idea to use consistent Bootstrap styling (https://getbootstrap.com/docs/5.2/forms/overview/) with proper layout management instead of `<br />` etc.

File: app/views/landing_page/index.html.erb
Comment (luap42) [inline app/views/landing_page/index.html.erb:21]: Suggestion: put this link into line 12, so that clicking on your own email address leads you to your profile

Comment (luap42) [reply app/views/landing_page/index.html.erb:21]: Other suggestion: instead of showing just the email, what about showing both the email and the full name? (only showing the full name is probably not a good idea because it is not - yet - guaranteed to be set immediately after signup, which means that the link would be empty and unclickable otherwise)

My idea would be to have something like "Test User (test@example.org)".

File: app/views/profiles/show.html.erb
Comment (luap42) [inline app/views/profiles/show.html.erb:4]: Apparently there is a [`mail_to` helper](https://apidock.com/rails/ActionView/Helpers/UrlHelper/mail_to) that would generate this code for us; this is IMO preferable because otherwise for example special characters in emails might cause trouble.

File: app/views/profiles/show.html.erb
Comment (luap42) [inline app/views/profiles/show.html.erb:8]: Nitpicky suggestion: instead of using a `<hr>` it might be preferable to have two `.card-body`s next to each other with a CSS top border, something like this:

```
<div class="card">
    <div class="card-body"><!-- meta info --></div>
    <div class="card-body border-top"><!-- description --></div>
</div>
```

File: config/routes.rb
Comment (luap42) [inline config/routes.rb:11]: Since `"profile"` is an identifier and not a mutable or complex string, it would be preferable to use `:profile`, i.e. Ruby's symbol type instead.

File: spec/features/users/show_spec.rb
Comment (luap42) [inline spec/features/users/show_spec.rb:28]: IMO there should be two tests here:

- User profile page lists the user's description (if it has one)
- User profile page doesn't list the user's description (if it has none)

File: spec/features/users/show_spec.rb
Comment (luap42) [inline spec/features/users/show_spec.rb:33]: In accordance with a review comment before I would suggest adding a test

- Going to `/profiles/me` redirects to profile URL with own profile ID

File: app/models/user.rb
Comment (luap42) [inline app/models/user.rb:8]: These two helper invocations should be superfluous, because Rails will automatically load the attributes from the DB Schema.

According to the official documentation (https://api.rubyonrails.org/classes/ActiveRecord/Attributes/ClassMethods.html#method-i-attribute) this method is for *overriding* this behaviour, which we do not want to do here, because the default behaviour should be just fine.

**Review by luap42 — CHANGES_REQUESTED** [review]
Looking good overall, but some changes are still needed.

File: app/controllers/profiles_controller.rb
Comment (Skn0tt) [reply app/controllers/profiles_controller.rb]: resolved in https://github.com/hpi-swt2/bookkeeper-portal-red/pull/44/commits/30f3e8fb4cc7dea36d12bb49d696821e5efb476b

File: app/controllers/profiles_controller.rb
Comment (Skn0tt) [reply app/controllers/profiles_controller.rb]: implemented in https://github.com/hpi-swt2/bookkeeper-portal-red/pull/44/commits/30f3e8fb4cc7dea36d12bb49d696821e5efb476b

File: app/views/devise/registrations/edit.html.erb
Comment (Skn0tt) [reply app/views/devise/registrations/edit.html.erb:34]: could you elaborate on what Bootstrap features we should use instead, and why they're superior to `<br/>`? We're new with bootstrap, so just trying to learn :)

File: app/views/landing_page/index.html.erb
Comment (Skn0tt) [reply app/views/landing_page/index.html.erb:21]: These are out-of-scope of the user stories. Let's circle back with @Ngo-Tuan-Kiet and get them prioritise for the next sprints.

File: app/views/profiles/show.html.erb
Comment (Skn0tt) [reply app/views/profiles/show.html.erb]: Today I learned about `mail_to`! Nice, implemented in https://github.com/hpi-swt2/bookkeeper-portal-red/pull/44/commits/30f3e8fb4cc7dea36d12bb49d696821e5efb476b

File: app/views/profiles/show.html.erb
Comment (Skn0tt) [reply app/views/profiles/show.html.erb:10]: Let's stick with the `<hr>`, if we find problems with it in the future we'll resort to two `.card-body` elements.

File: config/routes.rb
Comment (Skn0tt) [reply config/routes.rb:11]: I see the Ruby guidelines aren't consistent on this: https://guides.rubyonrails.org/routing.html#overriding-the-named-route-helpers, https://guides.rubyonrails.org/routing.html#naming-routes.

Could you elaborate on when to use one over the other, and what the tradeoffs are?

File: spec/features/users/show_spec.rb
Comment (Skn0tt) [reply spec/features/users/show_spec.rb:28]: keeping our approach towards test coverage in mind, let's stick with the tests we have and not over-do it :)

File: spec/features/users/show_spec.rb
Comment (Skn0tt) [reply spec/features/users/show_spec.rb:33]: We're thinking of the redirect as being an implementation detail. I think it's fine to assert on the outcome (what the page shows), not on how the browser gets there (redirect).

File: app/models/user.rb
Comment (Skn0tt) [reply app/models/user.rb:8]: We added these fields to have as documentation inside of the `user.rb` class, where we'll need to refer to them often. What do you think about having these lines as comments instead?

File: config/routes.rb
Comment (luap42) [reply config/routes.rb:11]: Hm... unsure about the tradeoffs, but I just looked at a list of open source projects using Ruby (went on https://github.com/asyraffff/Open-Source-Ruby-and-Rails-Apps and clicked on names that sounded familiar to me) and they all use symbols, which indicates for me that they are more Ruby-idiomatic:

- https://github.com/mastodon/mastodon/blob/main/config/routes.rb#L50
- https://github.com/forem/forem/blob/main/config/routes.rb#L15
- https://github.com/gitlabhq/gitlabhq/blob/master/config/routes.rb#L86
- https://github.com/discourse/discourse/blob/main/config/routes.rb#L160
- https://github.com/openstreetmap/openstreetmap-website/blob/master/config/routes.rb#L19

File: spec/features/users/show_spec.rb
Comment (luap42) [reply spec/features/users/show_spec.rb:28]: :+1:

File: spec/features/users/show_spec.rb
Comment (luap42) [reply spec/features/users/show_spec.rb:33]: :+1:

File: config/routes.rb
Comment (Skn0tt) [reply config/routes.rb:11]: Sounds good 👍
```suggestion
    get "/profiles/:id", to: "profiles#show", as: :profile
```

File: app/views/profiles/show.html.erb
Comment (Skn0tt) [reply app/views/profiles/show.html.erb]: even better! done in https://github.com/hpi-swt2/bookkeeper-portal-red/pull/44/commits/5887f503fbe0fe0bbf71ad11ee033016d5b2e14c

**Comment by Skn0tt** [comment]
we're getting merge conflicts left and right, let's get this merged! @luap42 could you give it another look and approve if you're fine with the contents? (let's bias towards getting this out fast + fixing later, I really don't want to deal with more merge conflicts ...)

**Comment by yannik-dittmar** [comment]
hey @luap42, you already did some great reviewing here. Could you maybe sign off the last changes, so the PR can be merged after resolving the small merge conflict?

**Review by JannisHajda — APPROVED** [review]

**Pull request merged**
